h3/src/frame/mod.rs

```diff
@@ -1,3 +1,284 @@
-#[allow(dead_code)]
-// TODO: remove once methods are effectively used through public API.
 mod buf;
+
+use std::{
+    io,
+    task::{Context, Poll},
+};
+
+use bytes::{Buf, Bytes};
+
+use crate::{
+    proto::{
+        frame::{self, Frame},
+        ErrorCode,
+    },
+    quic::RecvStream,
+};
+use buf::BufList;
+
+pub struct FrameStream<R: RecvStream> {
+    recv: R,
+    bufs: BufList<Bytes>,
+    decoder: FrameDecoder,
+    remaining_data: Option<u64>,
+}
+
+impl<R> FrameStream<R>
+where
+    R: RecvStream<Buf = Bytes>,
+{
+    pub fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Result<Option<Frame>, Error>> {
+        assert!(self.remaining_data.is_none());
+
+        let end = self.try_recv(cx)?;
+
+        match self.decoder.decode(&mut self.bufs)? {
+            Some(Frame::Data { len }) => {
+                self.remaining_data = Some(len);
+                Poll::Ready(Ok(Some(Frame::Data { len })))
+            }
+            Some(frame) => Poll::Ready(Ok(Some(frame))),
+            None if end => Poll::Ready(Ok(None)),
+            None => Poll::Pending,
+        }
+    }
+
+    pub fn poll_data(&mut self, cx: &mut Context<'_>) -> Poll<Result<Option<Bytes>, Error>> {
+        match self.remaining_data {
+            None => return Poll::Ready(Ok(None)),
+            Some(0) => {
+                self.remaining_data = None;
+                return Poll::Ready(Ok(None));
+            }
+            _ => (),
+        }
+
+        let end = self.try_recv(cx)?;
+
+        let data = self.bufs.bytes();
+        match (data.remaining(), end) {
+            (0, true) => return Poll::Ready(Ok(None)),
+            (0, false) => return Poll::Pending,
+            (x, _) => *self.remaining_data.as_mut().unwrap() -= x as u64,
+        };
+
+        Poll::Ready(Ok(Some(Bytes::from(data))))
+    }
+
+    pub fn reset(&mut self, error_code: ErrorCode) {
+        let _ = self.recv.stop_sending(error_code.0.into());
+    }
+
+    fn try_recv(&mut self, cx: &mut Context<'_>) -> Result<bool, Error> {
+        match self.recv.poll_data(cx) {
```

Comment:
shouldn't `recv` be polled continuously until returning `Poll::Pending` as to [to register interest](https://doc.rust-lang.org/nightly/core/task/enum.Poll.html#variant.Pending)?
 otherwise it seems there's a state where `revc.poll_data()` returns `Some(d))` but `FrameDecoder` returns `None` due to the frame being `Incomplete` and `FrameStream` returns `Poll::Pending`  without setting to be woken when `recv` makes progress . 
See the [h2 example](https://github.com/hyperium/h2/blob/master/src/codec/framed_read.rs#L347) on how to keep polling until a `Frame` can be decoded

Reply:
Looking closer into it, it looks like you're right! Thanks.

Reply:
I think I got this sorted out. Could you check again ?